Fix TypeError when duplicating uncategorized theme patterns

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What Why & How
<!-- In a few words, what is the PR actually doing? -->
When a pattern is duplicated, there is a modal with an input field that contains the labels of the pattern categories:
the information about the category is passed on to the new pattern.

The function that gets the labels compares the category from the theme pattern with the existing registered pattern categories, to make sure that the category exists and is displayed with the correct label.

However, the code did not take into account that theme patterns can be registered without the optional `categories` parameter.
This PR makes a small change to the filter and mapping using optional chaining, to avoid a TypeError when the `categories` parameter is not used.

Closes https://github.com/WordPress/gutenberg/issues/66888

## Testing Instructions
Activate a theme that registers a pattern without the `categories` parameter. This can be a classic theme or a block theme, for example Twenty Twenty-Five.
Open the Patterns screen from Appearance > Patterns or from Appearance > Editor > Patterns.
Locate the pattern that does not have a category.
Select the option to duplicate the pattern.
Confirm that the modal with the information about the duplicated pattern opens and that there are no JavaScript errors in the browser console.
Click the button at the bottom of the modal to duplicate the pattern. Confirm that there are no issues related to the category of the new pattern.

Now, select a different pattern that has a single category or multiple categories, and duplicate it.
Confirm that the category labels show correctly in the modal and that there are no new issues/ regressions.
Click the button at the bottom of the modal to duplicate the pattern. Confirm that there are no issues related to the category of the new pattern.

carolinan committed:
Fix TypeError when duplicating uncategorized patterns

Only try to match the pattern categories and get the label if the theme pattern
uses the `categories` parameter.

Otherwise there will be a TypeError when using the filter.

ntsekouras commented on packages/patterns/src/components/duplicate-pattern-modal.js:
```diff
@@ -15,11 +15,13 @@ import { PATTERN_SYNC_TYPES, PATTERN_TYPES } from '../constants';
+				?.filter( ( category ) =>
+					pattern.categories?.includes( category.name )
+				)
+				.map( ( category ) => category.label ) || []
```
Do we also need the empty array fallback? It seems it's handled as default by `CreatePatternModalContents`.

carolinan replied:
Thank you, I tested it again manually and removing the fallback does not seem to cause any issues.

ntsekouras reviewed (approved):
Left a small comment, but this looks good. Thank you!

carolinan committed:
Remove the empty array fallback

The array fallback is not needed inside getTermLabels, because getTermLabels is used as the value for defaultCategories, which is an empty array by default when it is used inside the function CreatePatternModalContents.

carolinan commented:
I briefly looked into adding a Playwright test for this but I was not able to figure out if there was a way to test a theme provided pattern.
Is there any way to tell the DataViews that the pattern origin is the theme, using `requestUtils.createBlock` ?

carolinan commented:
Oh, or I am supposed to get theme patterns by `await requestUtils.activateTheme( 'theme name here' );`

carolinan commented:
So I think the test should wait until TT5 is released, so that it can be added to the gutenberg-test-themes.

ntsekouras commented:
For tests we're using the empty theme, where we can add stuff, like templates, patterns etc.. Did you try adding one there?

carolinan commented:
I did not: I am torn about that because then it wont be _empty_ 😅

ntsekouras commented:
>then it wont be empty

I get it 😄 , but it's not entirely empty right now. Anyway.. This fix is a JS pretty strait forward one, and I believe it would be fine not to add a test.

mcsf commented:
Five lines below the patch site, there's a similar filter, but accessing property `wp_pattern_category` instead. Shouldn't we add optional chaining there too?

```diff
.filter( ( category ) =>
-	pattern.wp_pattern_category.includes( category.id )
+	pattern.wp_pattern_category?.includes( category.id )
)
```

@carolinan

carolinan commented:
@mcsf I left it because I was not able to reproduce any errors when testing with user created patterns.
Since it is an existing taxonomy, registered on the post type, I believe that if there are no categories, it would be an empty array?
But yes, _in theory_ there is a chance that someone would unregister the `wp_pattern_category` taxonomy from the `wp_block` post type.

carolinan commented:
@mcsf @ntsekouras 
Are there any other scenarios than the above where `wp_pattern_category` would not exist?

I ran a quick test and if I unregister the `wp_block_pattern` taxonomy, the block editor and site editor breaks because the taxonomy is used in [mapUserPattern ](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/private-selectors.js#L324)

The duplication of user created patterns do not break. I think it is something that needs to be looked at separately.

cbravobernal commented:
I just cherry-picked this PR to the wp/6.7 branch to get it included in the next release.
